bindings/nodejs/src/functions/oic-server.cc

```diff
@@ -0,0 +1,262 @@
+        .post = { .handle = defaultPost },
+        .del = { .handle = defaultDel }
+    };
+
+    ASSIGN_STR_SLICE_MEMBER_FROM_PROPERTY(local, js, error, resource_type);
+    ASSIGN_STR_SLICE_MEMBER_FROM_PROPERTY(local, js, error, interface);
+    ASSIGN_STR_SLICE_MEMBER_FROM_PROPERTY(local, js, error, path);
+
+    *definition = local;
+    return true;
+
+path_failed:
+    free((void *)(local.interface.data));
+interface_failed:
+    free((void *)(local.resource_type.data));
+resource_type_failed:
+    Nan::ThrowError(error);
+    return false;
+}
+
+NAN_METHOD(bind_sol_oic_server_add_resource) {
+    VALIDATE_ARGUMENT_COUNT(info, 2);
+    VALIDATE_ARGUMENT_TYPE(info, 0, IsObject);
+    VALIDATE_ARGUMENT_TYPE(info, 1, IsUint32);
+
+    struct sol_oic_server_resource *resource = 0;
+
+    struct sol_oic_resource_type resourceType;
+    if (!c_sol_oic_resource_type(Nan::To<Object>(info[0]).ToLocalChecked(),
+        &resourceType)) {
+        return;
+    }
+
+    struct ResourceInfo *resourceInfo =
+        ResourceInfo::New(Nan::To<Object>(info[0]).ToLocalChecked());
+    if (!resourceInfo) {
+        return;
+    }
+
+    resource = sol_oic_server_add_resource(&resourceType, resourceInfo,
+        (enum sol_oic_resource_flag)(Nan::To<uint32_t>(info[1]).FromJust()));
+    resourceInfo->resource = resource;
+
+    free((void *)(resourceType.resource_type.data));
+    free((void *)(resourceType.interface.data));
+    free((void *)(resourceType.path.data));
+
+    if (!resource) {
+        delete resourceInfo;
+    } else {
+        info.GetReturnValue().Set(SolOicServerResource::New(resourceInfo));
+    }
+}
+
+NAN_METHOD(bind_sol_oic_server_del_resource) {
+    VALIDATE_ARGUMENT_COUNT(info, 1);
+    VALIDATE_ARGUMENT_TYPE(info, 0, IsObject);
+    Local<Object> jsResourceInfo = Nan::To<Object>(info[0]).ToLocalChecked();
+    struct ResourceInfo *resourceInfo = (struct ResourceInfo *)
+        SolOicServerResource::Resolve(jsResourceInfo);
+    if (!resourceInfo) {
+        return;
+    }
+    sol_oic_server_del_resource(resourceInfo->resource);
+    delete resourceInfo;
+    Nan::SetInternalFieldPointer(jsResourceInfo, 0, 0);
+}
+
+NAN_METHOD(bind_sol_oic_notify_observers) {
+    VALIDATE_ARGUMENT_COUNT(info, 2);
+    VALIDATE_ARGUMENT_TYPE(info, 0, IsObject);
+    VALIDATE_ARGUMENT_TYPE_OR_NULL(info, 1, IsObject);
+    struct ResourceInfo *resourceInfo = (struct ResourceInfo *)
+        SolOicServerResource::Resolve(
+            Nan::To<Object>(info[0]).ToLocalChecked());
+    if (!resourceInfo) {
+        return;
+    }
+
+    Nan::Persistent<Object> *jsPayload = 0;
+    if (!info[1]->IsNull()) {
+        jsPayload = new Nan::Persistent<Object>(
+            Nan::To<Object>(info[1]).ToLocalChecked());
+    }
+
+    info.GetReturnValue().Set(Nan::New(
+        sol_oic_notify_observers(resourceInfo->resource,
+            (jsPayload ? oic_map_writer_callback : 0), jsPayload)));
+
+    if (jsPayload) {
+        jsPayload->Reset();
```

Comment:
is it necessary ? shouldn't the destructor be responsilbe to clear all the necessary data ?